framework/include/variables/MooseVariableFE.h

```diff
@@ -682,22 +686,93 @@ class MooseVariableFE : public MooseVariableField<OutputType>
+                      const StateArg & state) const override final;
+  DotType evaluateDot(const ElemArg &, const StateArg &) const override final;
+  DotType evaluateDot(const FaceArg &, const StateArg &) const override final;
+
+private:
+  /**
+   * Compute the solution, gradient, and time derivative with provided shape functions
+   */
+  template <typename Shapes, typename Solution, typename GradShapes, typename GradSolution>
+  void computeSolution(const Elem * elem,
+                       unsigned int n_qp,
+                       const StateArg & state,
+                       const Shapes & phi,
+                       Solution & local_soln,
+                       const GradShapes & grad_phi,
+                       GradSolution & grad_local_soln,
+                       Solution & dot_local_soln) const;
+
+  /**
+   * Evaluate solution and gradient for the \p elem_qp argument
+   */
+  void
+  evaluateOnElement(const ElemQpArg & elem_qp, const StateArg & state, bool cache_eligible) const;
+
+  /**
+   * Evaluate solution and gradient for the \p elem_side_qp argument
+   */
+  void evaluateOnElementSide(const ElemSideQpArg & elem_side_qp,
+                             const StateArg & state,
+                             bool cache_eligible) const;
+
+  /// Keep track of the current elem-qp functor element in order to enable local caching (e.g. if we
+  /// call evaluate on the same element, but just with a different quadrature point, we can return
+  /// previously computed results indexed at the different qp
+  mutable const Elem * _current_elem_qp_functor_elem = nullptr;
+
+  /// The values of the solution for the \p _current_elem_qp_functor_elem
+  mutable std::vector<ValueType> _current_elem_qp_functor_sln;
+
+  /// The values of the gradient for the \p _current_elem_qp_functor_elem
+  mutable std::vector<GradientType> _current_elem_qp_functor_gradient;
+
+  /// The values of the time derivative for the \p _current_elem_qp_functor_elem
+  mutable std::vector<DotType> _current_elem_qp_functor_dot;
+
+  /// Keep track of the current elem-side-qp functor element and side in order to enable local
+  /// caching (e.g. if we call evaluate with the same element and side, but just with a different
+  /// quadrature point, we can return previously computed results indexed at the different qp
+  mutable std::pair<const Elem *, unsigned int> _current_elem_side_qp_functor_elem_side{
+      nullptr, libMesh::invalid_uint};
+
+  /// The values of the solution for the \p _current_elem_side_qp_functor_elem_side
+  mutable std::vector<ValueType> _current_elem_side_qp_functor_sln;
+
+  /// The values of the gradient for the \p _current_elem_side_qp_functor_elem_side
+  mutable std::vector<GradientType> _current_elem_side_qp_functor_gradient;
+
+  /// The values of the time derivative for the \p _current_elem_side_qp_functor_elem_side
+  mutable std::vector<DotType> _current_elem_side_qp_functor_dot;
```

Comment:
I kind of wish there was a way to put this code in a separate class, so all the functor stuff lives together.

But a base class would need to still be in the current chain of variable classes to have all the attributes needed
and a derived class would be troublesome in its own way

Reply:
This kind of caching is unique to the reinit'ing of FE objects so to me this is an ideal place for it